segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/SegmentStateStore.java

```diff
@@ -72,40 +89,136 @@
+                .thenComposeAsync(sp -> {
+                    props[1] = sp;
+                    states[1] = readSegmentState(sp, timer.getRemaining());
+                    return states[1];
+                }, this.executor)
+                .exceptionally(this::handleSegmentNotExistsException);
+
+        return CompletableFuture.allOf(retVal1, retVal2).thenApplyAsync((v) -> {
+            SegmentState s1 = states[0] == null ? null : states[0].join();
+            SegmentState s2 = states[1] == null ? null : states[1].join();
+
+            if (s1 == null) {
+                return stateSegment1;
+            }
+            if (s2 == null) {
+                return stateSegment2;
+            }
+            if (props[0].getLastModified().asDate().compareTo(props[1].getLastModified().asDate()) < 0) {
+                return stateSegment1;
+            } else {
+                return stateSegment2;
+            }
+        }, this.executor);
+
     }
 
     @Override
     public CompletableFuture<Void> put(String segmentName, SegmentState state, Duration timeout) {
-        String stateSegment = StreamSegmentNameUtils.getStateSegmentName(segmentName);
         TimeoutTimer timer = new TimeoutTimer(timeout);
         ByteArraySegment toWrite = serialize(state);
 
         // We need to replace the contents of the Segment. The only way to do that with the Storage API is to
         // delete the existing segment (if any), then create a new one and write the contents to it.
-        return this.storage
-                .openWrite(stateSegment)
-                .thenComposeAsync(handle -> this.storage.delete(handle, timer.getRemaining()), this.executor)
-                .exceptionally(this::handleSegmentNotExistsException)
-                .thenComposeAsync(v -> this.storage.create(stateSegment, timer.getRemaining()), this.executor)
-                .thenComposeAsync(v -> this.storage.openWrite(stateSegment), this.executor)
-                .thenComposeAsync(
+        final String[] stateSegment = new String[1];
+        return this.findInvalidOrOlderState(segmentName, timeout)
+                   .thenComposeAsync( name -> {
+                       stateSegment[0] = name;
+                       return this.storage.openWrite(name);
+                   }, this.executor)
+                   .thenComposeAsync(handle -> this.storage.delete(handle, timer.getRemaining()), this.executor)
+                   .exceptionally(this::handleSegmentNotExistsException)
+                   .thenComposeAsync(v -> this.storage.create(stateSegment[0], timer.getRemaining()), this.executor)
+                   .thenComposeAsync(v -> this.storage.openWrite(stateSegment[0]), this.executor)
+                   .thenComposeAsync(
                         handle -> this.storage.write(handle, 0, toWrite.getReader(), toWrite.getLength(), timer.getRemaining()),
                         this.executor);
```

Comment:
We could delete the older file at the end.

Reply:
Added code the delete the file.